### PR TITLE
Explore different ranking models

### PR DESCRIPTION
**Pull request by rnyak**
This is addressing this open issue: https://github.com/NVIDIA-Merlin/models/issues/242

### CODE REVIEW

**Comment by marcromeyn** [comment]
I left some comments in ReviewNB.

File: examples/Exploring-different-models.ipynb [hunk 1]
```diff
@@ -0,0 +1,805 @@
...
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "a6e0eaaa",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "DATA_FOLDER = \"/workspace/data/\"\n",
+    "train_path = os.path.join(DATA_FOLDER, 'train/*.parquet')\n",
+    "test_path = os.path.join(DATA_FOLDER, 'test/*.parquet')\n",
+    "output_path = '/workspace/data/processed'"
```
Comment (bschifferer) [inline examples/Exploring-different-models.ipynb:116]: Should we use DATA_FOLDER as part of that path?

File: examples/Exploring-different-models.ipynb [hunk 2]
```diff
@@ -0,0 +1,805 @@
...
+   "id": "ebe79531",
+   "metadata": {},
+   "source": [
+    "NVTabular has already implemented multiple calculations, called `ops`. An `op` can be applied to a `ColumnGroup` from an overloaded `>>` operator\n",
+    "**Example:**<br>\n",
+    "```python\n",
+    "features = [ column_name, ...] >> op1 >> op2 >> ...\n",
+    "```\n",
+    "\n",
+    "This may sounds more complicated as it is. Let's define our the pipeline."
```
Comment (bschifferer) [inline examples/Exploring-different-models.ipynb:130]: Should we explain NVTabular + showing the NVTabular workflow.

1) We explain NVTabular in the NVTabular repo
2) We explain it in the 2nd notebook: Applying to your own dataset

Should we have a utils function for NVTabular workflow + returning merlin.io.Dataset ?
Should we use merlin.ioDataset instead of loading the schema + using the tf_dataloader.BatchedDataset

File: examples/Exploring-different-models.ipynb [hunk 3]
```diff
@@ -0,0 +1,805 @@
...
+    "### MLP Model"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "95ef3211",
+   "metadata": {},
+   "source": [
+    "<img src=\"./images/mlp.png\"  width=\"30%\">\n",
+    "We will first build a Multi-Layer Percepton (MLP) model. MLP models feed categorical features into embedding layer, concat the embedding outputs and add multiple hidden layers. Note that the `Ali-CCP` dataset has `click` and `conversion` target columns but in this example, we only focus on building different ranking models with binary target column `click`."
```
Comment (bschifferer) [inline examples/Exploring-different-models.ipynb:351]: Do we need a linebreak? It seems that there is one missing and the text starts after the image

File: examples/Exploring-different-models.ipynb [hunk 4]
```diff
@@ -0,0 +1,805 @@
...
+   "source": [
+    "save_results(\"DLRM\", model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c491573f",
+   "metadata": {},
+   "source": [
+    "### DCN Model"
```
Comment (bschifferer) [inline examples/Exploring-different-models.ipynb:659]: Which DCN model do we implement/train? The left or right?

File: examples/Exploring-different-models.ipynb (hunk 2 above)
Comment (rnyak) [reply examples/Exploring-different-models.ipynb:130]: Yes, we can remove NVTabular explanations. good point.

`Should we use merlin.ioDataset instead of loading the schema + using the tf_dataloader.BatchedDataset` --> this means we dump all to `model.fit()` and it makes model.fit() too crowded.. I prefer it as lean as existing tf.keras examples.

File: examples/Exploring-different-models.ipynb (hunk 3 above)
Comment (rnyak) [reply examples/Exploring-different-models.ipynb:351]: Yes, will fix it. thanks.

File: examples/Exploring-different-models.ipynb (hunk 4 above)
Comment (rnyak) [reply examples/Exploring-different-models.ipynb:659]: I will add. It is stacked by default. Thanks.

File: examples/example_utils.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+def save_results(model_name, model):
+    with open("results.txt", "a") as f:
+        f.write(model_name)
+        f.write("\n")
+        for key, value in model.history.history.items():
+            if "val_auc" in key:
+                f.write("%s:%s\n" % (key, value[0]))
+
+
+def create_bar_chart(text_file_name, models_name):
```
[the remark this answers is not in the document]
Comment (rnyak) [reply examples/example_utils.py:28]: I do not have time now to work on parsing the stored file. we can go with that simple solution.

File: examples/example_utils.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    workflow.transform(test_dataset).to_parquet(output_path=output_path + "/test/")
+
+
+def save_results(model_name, model):
+    with open("results.txt", "a") as f:
+        f.write(model_name)
+        f.write("\n")
+        for key, value in model.history.history.items():
+            if "val_auc" in key:
+                f.write("%s:%s\n" % (key, value[0]))
```
[the remark this answers is not in the document]
Comment (rnyak) [reply examples/example_utils.py:25]: we could :)

File: examples/Exploring-different-models.ipynb [hunk 7]
```diff
@@ -0,0 +1,805 @@
...
+    "\n",
+    "<a href=\"https://arxiv.org/pdf/1708.05031.pdf\">Image Source: NCF paper</a>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7b3ffb9b",
+   "metadata": {},
+   "source": [
+    "We will change the model to a Neural Collaborative Filtering (NCF). NCF feed categorical features into embedding layer, concat the embedding outputs and add multiple hidden layers via its MLP layer tower as seen in the figure. GMF and MLP share the same embedding layer, and then outputs of their interaction functions are concatenated.\n",
```
[the remark this answers is not in the document]
Comment (rnyak) [reply examples/Exploring-different-models.ipynb:464]: will check it out again.

File: examples/Exploring-different-models.ipynb (hunk 1 above)
Comment (rnyak) [reply examples/Exploring-different-models.ipynb:116]: Let's keep it as it is.

File: examples/Exploring-different-models.ipynb [hunk 8]
```diff
@@ -0,0 +1,802 @@
+{
```
Comment (gabrielspmoreira) [inline examples/Exploring-different-models.ipynb:1]: <p>When we use <code>Categorify(freq_threshold=5)</code>&nbsp; for user id and item id we end up with many examples with user_id = 0 and item_id=0. As the number of users/items in the long tail is usually large (about 43% of the Ali-CCP dataset), we might have the network learning wrong patterns (e.g. that every time item_id == 0 it is probably a click or not). We are not caring much about accuracy here but our users tend to use these examples as best practices for their own workflows.</p><p>I suggest including in the end of this workflow filtering out users and item ids = 0, like this</p><pre class="ql-syntax" spellcheck="false">outputs = outputs&gt;&gt; Filter(f=lambda df: (df["item_id"] != 0) &amp; (df["user_id"] != 0))
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/256/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (rnyak) [reply examples/Exploring-different-models.ipynb:1]: thanks @gabrielspmoreira. great point.

Comment (mikemckiernan) [inline examples/Exploring-different-models.ipynb:1]: <p>nit: I don't think the dataset name should be styled as inline code by the backticks. Plain text and initial caps ought to be fine.</p><p> </p><p>Do you mind replacing "here" as the link with an indication of the destination? Maybe "To download the training and test datasets, visit <a href="https://tianchi.aliyun.com/dataset/dataDetail?dataId=408#1" target="_blank">Ali-CCP: Alibaba Click and Conversion Prediction</a> at tianchi.aliyun.com."</p><p> </p><p>Maybe "We'll define several popular deep learning-based model architectures, train, and evaluate them and show how Merlin Models simplifies and eases this common and iterative process."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/256/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (mikemckiernan) [inline examples/Exploring-different-models.ipynb:1]: <p>Nice heading.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/256/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (mikemckiernan) [inline examples/Exploring-different-models.ipynb:1]: <p>Maybe this first para is just "First, import the libraries that we'll use in this notebook."</p><p> </p><p>(The para that introduces <code>import</code> statements is rarely a high information payload paragraph.)</p><p> </p><p>Or, if there's something to say about the use of TensorFlow, maybe add that.  I sincerely don't know, but I recall that models can work with Torch, so maybe you made a decision to use TensorFlow.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/256/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (mikemckiernan) [inline examples/Exploring-different-models.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #9.    from example_utils import workflow_fit_transform
</pre><p>I pitched "demo" instead of <code>merlin.models.data</code> when I reviewed Benedikt's notebook.  If the team can settle on either <code>demo</code> or <code>example</code>, it'd be a nice consistency.  I favor <code>demo_utils</code> FWIW.    </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/256/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (mikemckiernan) [inline examples/Exploring-different-models.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #19.    from merlin.schema.io.tensorflow_metadata import TensorflowMetadata
</pre><p>Nit that's not about the notebook, but the API.  If we have <code>merlin.models.tf</code>, then maybe we should have <code>merlin.schema.io.tf_metadata</code>.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/256/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (mikemckiernan) [inline examples/Exploring-different-models.ipynb:1]: <p>I don't think target encoding needs to be treated like a proper noun. Lower case should be fine. +100 for the brief explanation.</p><p> </p><p>+1 for mentioning a characteristic of the dataset--that it has categorical features only.  That's valuable to readers who are not familiar with the data.</p><p> </p><p>Please add a comma after "For example". In the same sentence, do you mind replacing "it" with "target encoding?" (Restating the noun instead of a pronoun (it, that) is tedious, but very clear.)</p><p> </p><p>Sugg to replace "here" link with the article title and domain, maybe "To learn more about TE, see <a href="https://medium.com/rapids-ai/target-encoding-with-rapids-cuml-do-more-with-your-categorical-data-8c762c79e784" target="_blank">Target Encoding with RAPIDS cuML: Do More with Your Categorical Data</a> at medium.com."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/256/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (mikemckiernan) [inline examples/Exploring-different-models.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #19.    outputs = user_id+item_id+targets+add_feat+te_feat
</pre><p>nit: Do you mind adding a space on either side of the <code>+</code> operator? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/256/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (mikemckiernan) [inline examples/Exploring-different-models.ipynb:1]: <p>If you're open to suggestion, please consider repeating the function name.</p><p> </p><p>"We use a utility function, <code>workflow_fit_transform</code> to perform workflow transformation. The function enables us to focus on defining, training, and evaluating models." </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/256/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (mikemckiernan) [inline examples/Exploring-different-models.ipynb:1]: <p>After reading the following paras and the cells--and that's golden content--taking the time to...</p><ul><li>mention the schema</li><li>how the schema was created</li><li>how models uses the schema</li><li>showing <code>schema.column_names</code> (so it is less magical)</li></ul><p>...those steps are all very valuable to anyone who is not already familiar with the schema object.</p><p> </p><p>So, I think the heading should change, but I'm not sure what to suggest.  The possible words to use are "NVTabular", "Schema", "Standard".</p><p> </p><p>Maybe "Benefits of Preparing Data with NVTabular" or "NVTabular and a Data Schema: Benefits of Standardization"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/256/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (mikemckiernan) [inline examples/Exploring-different-models.ipynb:1]: <p>If you can indicate what method or function caused NVT to export the schema, please do.</p><p> </p><p>There might be some unnecessary repetition in the second sentence. Maybe:</p><p> </p><p>&lt;fiction-warning please-check="true"&gt;</p><p>"The schema object identifies the input features. The Merlin Models library uses the schema to automatically build the layers to represent, normalize, and aggregate the input features."</p><p>&lt;/fiction-warning&gt;</p><p> </p><p>Sugg: lower case schema and init cap Merlin Models.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA-Merlin/models/pull/256/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (rnyak) [reply examples/Exploring-different-models.ipynb:1]: @mikemckiernan thanks. You can mention that in our models slack channel. But this will affect many code :)

Comment (rnyak) [reply examples/Exploring-different-models.ipynb:1]: @mikemckiernan thanks.

Comment (rnyak) [reply examples/Exploring-different-models.ipynb:1]: <p>Thanks. I will rephrase. </p>

Comment (rnyak) [reply examples/Exploring-different-models.ipynb:1]: <p>Sorry, not sure I understand what heading should change? <code>Training Recommender Models</code> ? We decided that with Even. I dont think excluding training rec models  from the  heading will be informative and correct.  It is what we do in the following cells: we are training the models with Merlin Models. May be I can change it as <code>Training Recommender Models with Merlin Models</code>.</p>

Comment (rnyak) [reply examples/Exploring-different-models.ipynb:1]: <p>thanks.</p>

Comment (rnyak) [reply examples/Exploring-different-models.ipynb:1]: <p>thanks :)</p>
